README.md

```diff
@@ -55,6 +55,55 @@ test.all(common)
 
 A valid (read: that follows this interface) IPFS core implementation, must expose the following API.
 
+## Files
+
+### `add`
+
+> Add files and data to IPFS.
+
+##### `Go` **WIP**
+
+##### `JavaScript` - ipfs.files.add(data, [callback])
+
+Where `data` may be
+
+- an array of objects, each of the form
+```js
+{
+  path: '/tmp/myfile.txt',
+  stream: fs.createReadStream('/tmp/myfile.txt')
+}
+```
+- an array of objects, each of the form
+```js
+{
+  path: '/tmp/myfile.txt',
+  content: new Buffer('hello')
+}
+```
+- a `Buffer` instance
+- a readable stream
+
+If no value `data` is provided, a Readable stream will be returned, which
+tuples like the above two object formats can be written.
+
+`callback` must follow `function (err, res) {}` signature, where `err` is an
+error if the operation was not successful and `res` is an array of objects of
+the form
+
+```js
+// js-ipfs-api
+{
+  Name: 'some name',
+  Hash: 'QmRNjDeKStKGTQXnJ2NFqeQ9oW23WcpbmvCVrpDHgDg3T6'
+}
+```
```

Comment:
The 'high cost' is virtual, because currently it is a given fact that the http-api needs improvements and it is better to provide a richer and consistent interface at a cost of another RTT, feeding back that into the http-api design, that simply cut it off.